### ISSUE TITLE
ALX: a program that prints the opcodes of its own main function.

### ISSUE THREAD

**Issue by ahmed149** [ahmed149/learning-programing-in-C#71]
#include <stdio.h>
#include <stdlib.h>

/**
 * main - function to print its own opcode
 * @argc: number of arguments
 * @argv: argument variables passed
 * Return: Always 0/
 */
int main(int argc, char **argv)
{
	int a, b;
	char *s;

	if (argc < 2)
	{
		printf("Error\n");
		exit(1);
	}
	a = atoi(argv[1]);
	if (a < 0)
	{
		printf("Error\n");
		exit(2);
	}
	s = (char *)main;
	for (b = 0; b < a - 1; b++)
		printf("%02hhx ", s[b]);
	printf("%02hhx\n", s[b]);
	return (0);
}

**Comment by ahmed149** [comment]
This C program is designed to print the opcodes (machine code) of its own instructions. Let's break down the code step by step:

1. `#include <stdio.h>` and `#include <stdlib.h>`:
   - These lines include standard C libraries for input/output and memory management, respectively.

2. `int main(int argc, char **argv)`:
   - This is the `main` function, which is the entry point of the program. It takes two parameters:
     - `argc`: The number of command-line arguments.
     - `argv`: An array of strings representing the command-line arguments.

3. `int a, b;` and `char *s;`:
   - These lines declare integer variables `a` and `b`, as well as a character pointer `s` for later use.

4. `if (argc < 2)`:
   - This condition checks if there are fewer than 2 command-line arguments, meaning that the user did not provide a number as an argument. If so, it prints "Error" and exits with a status code of 1 using `exit(1)`.

5. `a = atoi(argv[1]);`:
   - This line converts the first command-line argument, `argv[1]`, to an integer using the `atoi` function and stores it in the variable `a`.

6. `if (a < 0)`:
   - This condition checks if the integer `a` is less than 0, meaning it's a negative number. If so, it prints "Error" and exits with a status code of 2 using `exit(2)`.

7. `s = (char *)main;`:
   - This line assigns the address of the `main` function to the character pointer `s`. This is the key part of the program that allows it to examine its own machine code.

8. `for (b = 0; b < a - 1; b++)`:
   - This `for` loop iterates from 0 to `a - 1`, where `a` is the user-provided number. It's used to traverse the machine code instructions.

9. `printf("%02hhx ", s[b]);`:
   - Inside the loop, it prints the machine code at each position pointed to by `s`. The format specifier `%02hhx` is used to print each byte of machine code as a hexadecimal value with at least two digits. It also prints a space after each byte.

10. `printf("%02hhx\n", s[b]);`:
    - After the loop, this line prints the last byte of machine code followed by a newline character.

11. `return (0);`:
    - Finally, the `main` function returns 0 to indicate successful execution.

In summary, this program takes a positive integer as a command-line argument and then inspects its own machine code instructions to print them out in hexadecimal format. It performs error checks to ensure valid input and provides corresponding error messages if any issues are encountered.
